### PR TITLE
Add datasets

### PR DESCRIPTION
**Pull request by worksewell**
## Storage Providers

```
$ gradient storageProviders
Usage: gradient storageProviders [OPTIONS] COMMAND [ARGS]...

  Manage storage providers

Options:
  --help  Show this message and exit.

Commands:
  create   Create storage providers
  delete   Delete storage provider
  details  Show storage provider details
  list     List storage providers
  update   Update storage providers

$ gradient storageProviders create s3 --name test --bucket my-bucket --accessKey=access-key --secretAccessKey=secret-key
Created storage provider: splgct3arqdh77c

$ gradient storageProviders list
+-------+-----------------+------+---------------------------+
| Name  | ID              | Type | Config                    |
+-------+-----------------+------+---------------------------+
| test  | splgct3arqdh77c | s3   | bucket: my-bucket         |
|       |                 |      | accessKey: access-key     |
|       |                 |      | secretAccessKey: ******** |
+-------+-----------------+------+---------------------------+

$ gradient storageProviders details --id=splgct3arqdh77c
+--------+---------------------------+
| Name   | test                      |
+--------+---------------------------+
| ID     | splgct3arqdh77c           |
| Type   | s3                        |
| Config | bucket: my-bucket         |
|        | accessKey: access-key     |
|        | secretAccessKey: ******** |
+--------+---------------------------+

$ gradient storageProviders update s3 --id=splgct3arqdh77c --bucket=my-bucket3
Updated storage provider: splgct3arqdh77c

$ gradient storageProviders delete --id=splgct3arqdh77c
Deleted storage provider: splgct3arqdh77c
```

## Datasets

```
$ gradient datasets
Usage: gradient datasets [OPTIONS] COMMAND [ARGS]...

  Manage datasets

Options:
  --help  Show this message and exit.

Commands:
  create    Create dataset
  delete    Delete dataset
  details   Show dataset details
  files     Manage files
  list      List datasets
  tags      Manage tags
  update    Update dataset
  versions  Manage versions

$ gradient datasets create --name=test --storageProviderId=spltautet072md4
Created dataset: dst364npcw6ccok

$ gradient datasets list
+------+-----------------+-------------------------+
| Name | ID              | Storage Provider        |
+------+-----------------+-------------------------+
| test | dst364npcw6ccok | test1 (splgct3arqdh77c) |
+------+-----------------+-------------------------+

$ gradient datasets details --id=dst364npcw6ccok
+-----------------+-------------------------+
| Name            | test                    |
+-----------------+-------------------------+
| ID              | dst364npcw6ccok         |
| Description     |                         |
| StorageProvider | test1 (splgct3arqdh77c) |
+-----------------+-------------------------+

$ gradient datasets update --id=dst364npcw6ccok --description='Some description'
Updated dataset: dst364npcw6ccok

$ gradient datasets delete --id=dst364npcw6ccok
Deleted dataset: dst364npcw6ccok
```

## Dataset versions

```
$ gradient datasets versions
Usage: gradient datasets versions [OPTIONS] COMMAND [ARGS]...

  Manage versions

Options:
  --help  Show this message and exit.

Commands:
  commit   Commit dataset version
  create   Create dataset version
  delete   Delete dataset version
  details  Show dataset version details
  list     List dataset versions
  update   Update dataset version

$ gradient datasets versions create --id=dst364npcw6ccok --source-path=./some-data/
Created dataset version: dst364npcw6ccok:fo5rp4m
Committed dataset version: dst364npcw6ccok:fo5rp4m

$ gradient datasets versions list --id=dst364npcw6ccok
+-------------------------+---------+-------------+
| ID                      | Message | Tags        |
+-------------------------+---------+-------------+
| dst364npcw6ccok:fo5rp4m |         | test, test2 |
+-------------------------+---------+-------------+

$ gradient datasets versions details --id=dst364npcw6ccok:fo5rp4m
+-----------+-------------------------+
| ID        | dst364npcw6ccok:fo5rp4m |
+-----------+-------------------------+
| Message   |                         |
| Committed | true                    |
| Tags      | test, test2             |
+-----------+-------------------------+

$ gradient datasets versions update --id=dst364npcw6ccok:fo5rp4m --message='Some message'
Updated dataset version: dst364npcw6ccok:fo5rp4m

$ gradient datasets versions delete --id=dst364npcw6ccok:fo5rp4m
Deleted dataset version: dst364npcw6ccok:fo5rp4m
```

## Dataset tags

```
$ gradient datasets tags
Usage: gradient datasets tags [OPTIONS] COMMAND [ARGS]...

  Manage tags

Options:
  --help  Show this message and exit.

Commands:
  delete  Delete dataset tag
  list    List dataset tags
  set     Set dataset tag

$ gradient datasets tags set --id=dst364npcw6ccok:fo5rp4m --name=test
Set dataset tag: dst364npcw6ccok:test

$ gradient datasets tags list --id=dst364npcw6ccok
+-------------------------+------+
| ID                      | Name |
+-------------------------+------+
| dst364npcw6ccok:fo5rp4m | test |
+-------------------------+------+

$ gradient datasets tags delete --id=dst364npcw6ccok:test
Deleted dataset tag: dst364npcw6ccok:test
```

## Dataset files

```
$ gradient datasets files
Usage: gradient datasets files [OPTIONS] COMMAND [ARGS]...

  Manage files

Options:
  --help  Show this message and exit.

Commands:
  delete  Delete files
  get     Get files
  list    List files
  put     Put files

$ gradient datasets files list --id=dst364npcw6ccok:fo5rp4m
+-----------+------+
| Name      | Size |
+-----------+------+
| hello.txt | 12   |
+-----------+------+

$ gradient datasets files get --id=dst364npcw6ccok:fo5rp4m --target-path=output

# only works with uncommitted versions
$ gradient datasets files put --id=dst364npcw6ccok:xzxc3oo --source-path=./some-data/

# only works with uncommitted versions
$ gradient datasets files delete --id=dst364npcw6ccok:xzxc3oo
```

## Jobs

```
$ gradient jobs create \
  --clusterId=$CLUSTER \
  --machineType=$MACHINE_TYPE \
  --projectId=$PROJECT \
  --container=bash \
  --command='cat /datasets/input/hello.txt > /datasets/output/hello2.txt && date >> /datasets/output/hello2.txt' \
  --dataset=output@dst364npcw6ccok \
  --dataset=input@dst364npcw6ccok:fo5rp4m
```

### CODE REVIEW

File: gradient/api_sdk/clients/dataset_client.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        :rtype: list[models.Dataset]
+        """
+
+        repository = self.build_repository(repositories.ListDatasets)
+        return repository.list(limit=limit, offset=offset)
+
+    def delete(self, dataset_id):
+        """Delete a dataset
+
+        :param int dataset_id: ID [required]
```
Comment (worksewell) [inline gradient/api_sdk/clients/dataset_client.py:23]: `:param str`

File: gradient/api_sdk/clients/dataset_client.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        :rtype: None
+        """
+
+        repository = self.build_repository(repositories.DeleteDataset)
+        repository.delete(dataset_id)
+
+    def get(self, dataset_id):
+        """Delete a dataset
+
+        :param int dataset_id: ID [required]
```
Comment (worksewell) [inline gradient/api_sdk/clients/dataset_client.py:35]: `:param str`

File: gradient/api_sdk/clients/dataset_client.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            name=name,
+            storage_provider_id=storage_provider_id,
+            description=description,
+        )
+
+        repository = self.build_repository(repositories.CreateDataset)
+        return repository.create(dataset)
+
+    def update(self, dataset_id, name=None, description=None):
+        """Update an existing S3 dataset
```
Comment (worksewell) [inline gradient/api_sdk/clients/dataset_client.py:64]: `Update an existing dataset`

File: gradient/api_sdk/clients/dataset_client.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+        :param str dataset_id: Dataset ID
```
Comment (worksewell) [inline gradient/api_sdk/clients/dataset_client.py:66]: `[required]`

**Review by paperspace-philip — APPROVED** [review]

**Pull request merged**